src/fu-engine.c

```diff
@@ -1355,8 +1356,75 @@ fu_engine_check_requirement_hardware (FuEngine *self, XbNode *req, GError **erro
 	return FALSE;
 }
 
+static FwupdFeatureFlags
+fu_engine_get_features_for_sender (FuEngine *self, const gchar *sender)
+{
+	FwupdFeatureFlags *tmp;
+
+	g_return_val_if_fail (FU_IS_ENGINE (self), FWUPD_FEATURE_FLAG_LAST);
+	g_return_val_if_fail (sender != NULL, FWUPD_FEATURE_FLAG_LAST);
+
+	tmp = g_hash_table_lookup (self->sender_features, sender);
+	if (tmp == NULL)
+		return FWUPD_FEATURE_FLAG_NONE;
+	return *tmp;
+}
+
+static gboolean
+fu_engine_check_requirement_client (FuEngine *self,
+				    XbNode *req,
+				    const gchar *sender,
+				    GError **error)
+{
+	FwupdFeatureFlags flags;
+	g_auto(GStrv) feature_split = NULL;
+
+	/* should not be set in normal operation */
+	if (g_strcmp0 (sender, FU_ENGINE_DBUS_SENDER_INVALID) == 0) {
```

Comment:
Isn't this going to lock out a `fwupdtool` client that can show text messages?

Reply:
Yes, good catch. Either this `if` can just be deleted, or we add something like `FU_TOOL_SENDER_FAKE = "fake"`

Reply:
Well see my comment about a client setting features to the engine, I think the other option is to have fu-tool set features instead of this sender checking stuff.  You just check if the engine's current client supports the features, fu-tool included.

Reply:
> the engine's current client

Reply:
> the engine's current client

That seems a huge assumption for a shared daemon.

Reply:
But if you bundle `fwupd_client_install` with something that sets up all the current features?

Reply:
you need it way before install time; to check the requirements you need the featureset at `GetDetails` and `GetUpgrades` time for instance. There's no point showing the user they have an update with no indication it might not work but then just failing on the install step.

Reply:
But so bundle it with each call to the daemon.  When you do fwupd_client_get_details, update the features, you get the relevant results.  When you do fwupd_client_install, you update the features again.  I just think sending the sender id all over the daemon seems like a lot of unnecessary plumbing here.

Reply:
> But so bundle it with each call to the daemon

One thing I considered is breaking all the D-Bus method API and adding a `a{sv}` called `options` or something where each request could set the feature flags, the locale etc. That's probably architecturally cleanest, but it's a **huge** API break I'm super keen on right now -- a few projects are using the D-Bus APIs rather than the GObject client. Although translated device descriptions and release notes sounds like a good idea, it's impossible to get the content (both from the hardware, and from OEMs) and so I kinda gave up on that a while ago.

> I just think sending the sender id all over the daemon seems like a lot of unnecessary plumbing here.

What about rather than sending the `sender` all over the place we instead looked up the feature flags in the `fu-main.c` file and passed the feature_flags to each engine method. That means FU_ENGINE_DBUS_SENDER_INVALID kinda becomes unnecessary for fwupdtool.

> why not just add a new private detail for client_features

Reply:
>One thing I considered is breaking all the D-Bus method API 

But it's all abstracted by the library anyway.  And it re-inforces the point that people shouldn't be (ab)using the dbus API directly, they **should** be using the library.

> I'm super keen on right now

You mean _not_ super keen on presumably?  You seems to have conflicting thoughts.

>translated device descriptions and release notes sounds like a good idea, it's impossible to get the content (both from the hardware, and from OEMs) and so I kinda gave up on that a while ago.

You really need an army of translator monkeys for this.  I agree it's not worth the effort.

>fu_engine_set_feature_flags

Yes that's exactly what I was trying to convey.

Reply:
> they should be using the library.

It's KDE that wasn't keen on GObject from memory.

> You mean not super keen on presumably

Yes, sorry. As in I really don't want to break the world with a RHEL head on.

> fu_engine_set_feature_flags

I'll prototype something and see what happens.

Reply:
> I'll prototype something and see what happens.

So I started doing that, and it doesn't feel right. For instance, if you call into the daemon with fwupdmgr and then the daemon calls fu_engine_update_motd() then we use the features of the client to calculate the MOTD. If we don't call into the daemon first, then we have no features set and the MOTD is different. I think it's much better to be explicit for things like this otherwise it all gets a bit Rube Goldberg.

That said, I don't like passing the sender around, and the fact that FU_ENGINE_DBUS_SENDER_INVALID exists kinda indicates it's the wrong approach. The other thing I'd like to try is to pass around the FwupdFeatureFlags with the request -- or maybe wrap that up in a GObject something like FuRequest. The latter would let us add other things in the future without breaking internal ABI yet again.

Reply:
> maybe wrap that up in a GObject something like FuRequest

Also, this allows us to nicely tidy up the GetDevices() device flags which depends on the sender; the "if you're root you get the instance ID" thing.

Reply:
> MOTD is different

Hmm, I hadn't even started to consider the implications for the motd.  I almost think that should always be calculated as though all the features are supported.  The logic being that the motd message **tells you** to use `fwupdmgr`.

Reply:
> Hmm, I hadn't even started to consider the implications for the motd

Right, when making all the changes I kept having to think "what is set before this API call" which was a mental load that's ripe for getting wrong.

> motd message tells you to use fwupdmgr

Yes, we'd juts hardcode a FuEngineRequest with the same FwupdFeatureFlags as fwupdmgr.